### PR TITLE
COPY FROM should fail gracefully in case of illegal data in CSV

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Resolves https://github.com/crate/crate/issues/12201

## Checklist

 - [ ] Added an entry in `CHANGES.txt` for user facing changes
 - [ ] Updated documentation & `sql_features` table for user facing changes
 - [ ] Touched code is covered by tests
 - [ ] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [ ] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/main/java/io/crate/execution/engine/indexing/GroupRowsByShard.java [hunk 1]
```diff
@@ -125,18 +125,30 @@ public void accept(ShardedRequests<TReq, TItem> shardedRequests, Row row) {
...
-                shardedRequests.add(item, sizeEstimate, indexName, routing, rowSourceInfo);
-            } else {
-                shardedRequests.add(item, sizeEstimate, shardLocation, rowSourceInfo);
+
+            if (item instanceof ShardUpsertRequest.Item upsertRequestItem
+                &&
+                upsertRequestItem.insertValues() != null
+                &&
+                upsertRequestItem.insertValues().length > 0
+                && upsertRequestItem.insertValues()[0] != null) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/indexing/GroupRowsByShard.java:134]: I think `GroupRowsByShard` shouldn't be aware of concrete `ShardRequest.Item` implementations as it would break encapsulation. 

You could add a new property to the base class that tells consumers if the item succeeded or if it should be skipped.

File: server/src/main/java/io/crate/execution/engine/indexing/ShardedRequests.java [hunk 2]
```diff
@@ -81,16 +78,11 @@ public void add(TItem item, long itemSizeInBytes, ShardLocation shardLocation, R
     }
 
     void addFailedItem(String sourceUri, String readFailure, Long lineNumber) {
-        List<ReadFailureAndLineNumber> itemsWithFailure = itemsWithFailureBySourceUri.computeIfAbsent(
-            sourceUri, k -> new ArrayList<>());
+        Set<ReadFailureAndLineNumber> itemsWithFailure = itemsWithFailureBySourceUri.computeIfAbsent(
+            sourceUri, k -> new HashSet<>());
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/indexing/ShardedRequests.java:82]: Why the change here? How can it happen that there are multiple duplicate errors within the same line?

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/engine/indexing/ShardedRequests.java:82]: That's basically my current issue/question - details in the [comment](https://github.com/crate/crate/pull/12229#discussion_r834345142) above. In debugger` sourceUriFailureInput.value();` holds value of the previous iteration when parsing second line - I just came up with another idea how to "reset input" - would have to use `lineContext.setFailure(null) `after each process

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/engine/indexing/ShardedRequests.java]: https://github.com/crate/crate/pull/12229/commits/af7739f84577ad44b68b35c6268e906da0dbb461

reverted Set and used another approach to avoid side effects of the prev iteration - now duplicate error is resolved but success count is 0 - `setRawSource(null)` in catch block probably wrong

**Comment by BaurzhanSakhariev** [comment]
Added 4.7 label as I [reproduced](https://github.com/crate/crate/pull/12256) it even on 4.6

File: server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java [hunk 3]
```diff
@@ -1045,4 +1045,32 @@ public void testCopyFromWithValidationSetToFalseIgnoreCheckConstraints() throws
...
+
+    @Test
+    public void test_copy_from_fail_gracefully_in_case_of_invalid_data() throws Exception {
+        execute("create table t (obj object(dynamic) as (x int))");
+
+        // Checking out all variants from https://github.com/crate/crate/issues/12201#issuecomment-1072472337
+
+        List<String> lines = List.of(
+          "obj\n",
+            "\"{\"\"x\"\":1}\"\n",  // "{""x"":1}" - works
```
[the remark this answers is not in the document]
Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java]: done in https://github.com/crate/crate/commit/af7739f84577ad44b68b35c6268e906da0dbb461 and https://github.com/crate/crate/commit/4d2131bba90aa961bdc6d74c93d4a9ffaec14fce

File: server/src/main/java/io/crate/execution/engine/indexing/GroupRowsByShard.java (hunk 1 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/engine/indexing/GroupRowsByShard.java]: done in https://github.com/crate/crate/pull/12229/commits/6ea66a1780aca2f43bb4ba1f0073b45210436ca2

File: server/src/main/java/io/crate/execution/engine/indexing/GroupRowsByShard.java [hunk 4]
```diff
@@ -112,31 +112,38 @@ public void accept(ShardedRequests<TReq, TItem> shardedRequests, Row row) {
...
-            return;
-        }
 
-        try {
-            rowShardResolver.setNextRow(row);
-            for (int i = 0; i < expressions.size(); i++) {
-                expressions.get(i).setNextRow(row);
-            }
+        // Add uri failure to the summary if there exist any.
+        uriFailureRecorder.accept(shardedRequests);
```
[the remark this answers is not in the document]
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/engine/indexing/GroupRowsByShard.java:117]: it was like this, but as item is not always `ShardUpsertRequest` I had to do `instance of`and changed it in https://github.com/crate/crate/pull/12229/commits/6ea66a1780aca2f43bb4ba1f0073b45210436ca2. See also https://github.com/crate/crate/pull/12229#discussion_r834384492

File: server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java [hunk 5]
```diff
@@ -71,6 +74,7 @@ public SourceIndexWriterReturnSummaryProjection(RelationName relationName,
         sourceUri = (InputColumn) Symbols.fromStream(in);
         sourceUriFailure = (InputColumn) Symbols.fromStream(in);
         lineNumber = (InputColumn) Symbols.fromStream(in);
+        sourceParsingFailure = (InputColumn) Symbols.fromStream(in); //TODO this is draft, add version check
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java:77]: As per comment, need to check version here.

Not sure how to proceed - just retuning null `if version < 4.7.2` would make it the only nullable `Input`here and would have to check [null](https://github.com/crate/crate/pull/12229/files#diff-1931720ac30a5f869ae7f2449bd5bd9956dbfc9581bc3b6fb8d6aae3e9ce2c0bR146) here.

File: server/src/main/java/io/crate/execution/engine/collect/files/LineProcessor.java [hunk 6]
```diff
@@ -56,11 +56,25 @@ void readFirstLine(URI currentUri, InputFormat inputFormat, BufferedReader curre
 
     public void process(String line) throws IOException {
         lineContext.incrementCurrentLineNumber();
+        lineContext.setCurrentParsingFailure(null); // Reset prev failure if there is any.
```
Comment (jeeminso) [inline server/src/main/java/io/crate/execution/engine/collect/files/LineProcessor.java:59]: we could have a `reset` method similar to `resetCurrentLineNumber`

File: server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java (hunk 5 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java]: As you tagged it as a `bug`, I think `sourceParsingFailure` should be back ported. In which case you do not need version checks for streaming. 
In case where the version is lower that it is not back ported,
```
        } catch (IOException e) {
             lineProcessor.setFailure(e.getMessage());
             lineProcessor.setUriFailure(e.getMessage());
             return true;
         } catch (Exception e) {
             lineProcessor.setParsingFailure(e.getMessage());
             return true;
```

this catch blocks will also not be there and will just fail as a bug (won't reach the point, https://github.com/crate/crate/pull/12229/files#diff-53c11b5d99dc6cd6d5ea9e7a33a2f3f197c476ddd9675737246257aedbce2468R259).

Comment (jeeminso) [reply server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java]: > As per comment, need to check version here. Not sure how to proceed - just retuning [null](https://github.com/crate/crate/pull/12229/files#diff-1931720ac30a5f869ae7f2449bd5bd9956dbfc9581bc3b6fb8d6aae3e9ce2c0bR146) if version < 4.7.2 would make it the only nullable Inputhere and would have to check null here.

This is resolved by your new commit https://github.com/crate/crate/pull/12229/commits/aaa55ce7979d0742fbd6b189ee652638ea8d5121 right?

Comment (jeeminso) [reply server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java]: > Currently implemented by [creating new instance](https://github.com/crate/crate/pull/12229/commits/aaa55ce7979d0742fbd6b189ee652638ea8d5121)

It looks like
```
toCollect.add(SourceParsingFailureExpression.getReferenceForRelation(table.ident())); // CopyFromPlan
```
is streamed through `FileUriCollectPhase` as well. So if the ColumnIdent is recovered, `toCollect` should also be recovered(consistency wise, not that it has to in order to run successfully).

Comment (jeeminso) [reply server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java]: In my opinion, having the null check seems more preferable. Could you explain why you would `new` instead?

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java]: > In my opinion, having the null check seems more preferable. Could you explain why you would `new` instead?

Only style/consistency reasons - in order not to have mixed situation when some inputs are nullable and some are not in the `UpsertResultContext`. But on the other hand, if it legit can be null maybe it's actually better and less questionable. 

I'm currently inclined towards reverting second commit and make null check unless there is strong reason to make it non-nullable. Let's see what @mfussenegger says.

File: server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java [hunk 7]
```diff
@@ -71,6 +76,7 @@ public SourceIndexWriterReturnSummaryProjection(RelationName relationName,
         sourceUri = (InputColumn) Symbols.fromStream(in);
         sourceUriFailure = (InputColumn) Symbols.fromStream(in);
         lineNumber = (InputColumn) Symbols.fromStream(in);
+        sourceParsingFailure = in.getVersion().before(Version.V_4_7_1) ? new InputColumn(3, DataTypes.STRING) : (InputColumn) Symbols.fromStream(in);
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java:79]: I think the before 4.7.1 case deserves a comment. Why InputColumn 3?

File: server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java [hunk 8]
```diff
@@ -100,6 +110,7 @@ public void writeTo(StreamOutput out) throws IOException {
         Symbols.toStream(sourceUri, out);
         Symbols.toStream(sourceUriFailure, out);
         Symbols.toStream(lineNumber, out);
+        Symbols.toStream(sourceParsingFailure, out); // This field was introduced in 4.7.2
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java:113]: There needs to be a version check here too - could be sending it to an older node

File: server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java (hunk 7 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java:79]: > why 3

was not clear to me whether I should follow the way I create it [here](https://github.com/crate/crate/pull/12229/files#diff-53c11b5d99dc6cd6d5ea9e7a33a2f3f197c476ddd9675737246257aedbce2468R259) - probably it's better to simply return null (with null check in another place consuming this Input)? https://github.com/crate/crate/pull/12229#discussion_r839255280

File: server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java (hunk 8 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java]: https://github.com/crate/crate/pull/12229/commits/61d5b9ff921f756b13ceba46b52b85455392e214

File: server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java (hunk 7 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java:81]: I added a comment but haven't touched InputColumn - is streaming null a better idea? discussed [here](https://github.com/crate/crate/pull/12229#discussion_r839255280) with @jeeminso ?

**Review by mfussenegger — COMMENTED** [review]
From what I can tell this looks good - though I think @seut or @mkleen should have a quick look as well - as they're afaik more familiar with the UpsertResultContext/failure handling logic.

File: server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java (hunk 7 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/dsl/projection/SourceIndexWriterReturnSummaryProjection.java]: Good point, definitely need to add comments why 3. Also [implemented](https://github.com/crate/crate/pull/12229/commits/ac82a333e4a8e6682c1af600b76b006b6797582a) nullable approach - probably this one is slightly better as null is kinda self-documenting, at least no need to explain in 2 places pages of the specific index.

**Review by mfussenegger — APPROVED** [review]

**Review by seut — APPROVED** [review]
👍
